File: packages/renderer-web/src/states.ts
/**
 * How a node's runtime state collapses into one style — pure, so the precedence
 * order is unit-testable without a canvas and the Unity SDK has an unambiguous
 * reference for the same rule.
 *
 * There is no cascade: each node resolves its OWN base style plus the overrides
 * of the states it is currently in (decision 2026-08-04). The order below is the
 * whole contract, and it matters because states overlap — a pressed button is
 * usually also hovered and focused.
 */

import type { StateName, StateOverride, Style } from "@zabloo/format";

/** The runtime flags the renderer owns, keyed by component identity. */
interface NodeStates {
  hovered: boolean;
  pressed: boolean;
  focused: boolean;
  /** The chosen button of an `"exclusive-select"` group (a tab). */
  selected: boolean;
  /** A Toggle's own value (in an `"exclusive-check"` group, the group's). */
  checked: boolean;
  /** A TextInput holding no text — what styles its placeholder (ZAB-26). */
  empty: boolean;
  /**
   * Out of the interaction model, its own or an ancestor's (ZAB-63). The only
   * state a node that is not focusable can be in, which is what lets the labels
   * of a disabled section dim with the controls in it.
   */
  disabled: boolean;
}

/**
 * Least to most specific — later wins.
 *
 * Value states go first: what the control IS is the baseline, and how the
 * pointer or the focus is treating it right now paints over that. `empty` opens
 * them because it is the emptiest statement a control makes about its value (a
 * placeholder must lose to anything the author says about a selected or focused
 * field). `hover` sits under `focused` so a focus ring is never hidden by a mouse
 * passing by, and `pressed` wins over those because it lasts exactly as long as
 * the finger is down.
 *
 * `disabled` closes the list (ZAB-63). Its place only matters against the VALUE
 * states — a disabled node takes no input, so hover/focused/pressed can never be
 * active alongside it, while a disabled Toggle is still `checked` and a disabled
 * field still `empty`. Last is what lets one override speak for the whole control
 * whatever value it happens to hold.
 */
const STATE_ORDER: readonly StateName[] = [
  "empty",
  "selected",
  "checked",
  "hover",
  "focused",
  "pressed",
  "disabled",
];

function isActive(name: StateName, states: NodeStates): boolean {
  switch (name) {
    case "empty":
      return states.empty;
    case "selected":
      return states.selected;
    case "checked":
      return states.checked;
    case "hover":
      return states.hovered;
    case "focused":
      return states.focused;
    case "pressed":
      return states.pressed;
    case "disabled":
      return states.disabled;
    default:
      return false;
  }
}

/**
 * The style to resolve this frame: the base with every active state's override
 * merged over it, in `STATE_ORDER`. Returns the base itself when no state is
 * active, so an untouched node allocates nothing.
 */
function effectiveStyle(
  base: Style | undefined,
  // Keyed loosely, like the renderer reads the IR: an unknown state name is just
  // one this build never activates (forward tolerance), not a parse error.
  overrides: Readonly<Record<string, StateOverride | undefined>> | undefined,
  states: NodeStates,
): Style | undefined {
  if (!overrides) return base;
  let style = base;
  for (const name of STATE_ORDER) {
    const override = overrides[name]?.style;
    if (override && isActive(name, states)) style = { ...style, ...override };
  }
  return style;
}

export type { NodeStates };
export { effectiveStyle, STATE_ORDER };

File: core/src/states.cpp
#include <cstddef>
#include <optional>

#include "states.h"

namespace zabloo {

const StateName STATE_ORDER[static_cast<size_t>(StateName::Count)] = {
    StateName::Empty,   StateName::Selected, StateName::Checked,  StateName::Hover,
    StateName::Focused, StateName::Pressed,  StateName::Disabled,
};

namespace {

bool is_active(StateName name, const NodeStates &states) {
  switch (name) {
    case StateName::Empty: return states.empty;
    case StateName::Selected: return states.selected;
    case StateName::Checked: return states.checked;
    case StateName::Hover: return states.hovered;
    case StateName::Focused: return states.focused;
    case StateName::Pressed: return states.pressed;
    case StateName::Disabled: return states.disabled;
    case StateName::Count: break;
  }
  return false;
}

template <typename T>
void take(T &into, const T &from) {
  if (from.present()) into = from;
}

template <typename T>
void take_optional(std::optional<T> &into, const std::optional<T> &from) {
  if (from.has_value()) into = from;
}

void merge(Style &into, const Style &from) {
  take(into.background, from.background);
  take(into.border_color, from.border_color);
  take(into.color, from.color);
  take(into.radius, from.radius);
  take(into.border_width, from.border_width);
  take(into.font_size, from.font_size);
  take(into.line_height, from.line_height);
  take_optional(into.opacity, from.opacity);
  take_optional(into.max_lines, from.max_lines);
  take_optional(into.text_align, from.text_align);
  take_optional(into.text_align_y, from.text_align_y);
  take_optional(into.overflow, from.overflow);
  take_optional(into.wrap, from.wrap);
}

}  // namespace

Style effective_style(const Node &node, const NodeStates &states) {
  Style merged = node.style;
  for (const StateName name : STATE_ORDER) {
    const size_t index = static_cast<size_t>(name);
    if (!node.has_state[index] || !is_active(name, states)) continue;
    merge(merged, node.state_style[index]);
  }
  return merged;
}

}  // namespace zabloo
